### PR TITLE
[SYCLomatic] Support TemplateSpecializationTypeLoc rewriter

### PR DESCRIPTION
**Pull request by AndyCHHuang**

### CODE REVIEW

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -17357,6 +17357,31 @@ void ComplexAPIRule::runRule(
...
+                    loc(qualType(hasDeclaration(namedDecl(TargetTypeName())))))
+                    .bind("loc"),
+                this);
+}
+
+void TemplateSpecializationTypeLocRule::runRule(
+    const ast_matchers::MatchFinder::MatchResult &Result) {
+  if (auto TL = getNodeAsType<TypeLoc>(Result, "loc")) {
+    // Only need to apply the migration to ElaboratedType
+    if (auto ETL = TL->getAs<ElaboratedTypeLoc>()) {
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/ASTTraversal.cpp]: What if ETL is null?

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 2]
```diff
@@ -421,9 +425,9 @@ void ExprAnalysis::analyzeExpr(const DeclRefExpr *DRE) {
     addReplacement(DRE, TemplateDecl->getIndex());
   else if (auto ECD = dyn_cast<EnumConstantDecl>(DRE->getDecl())) {
     auto &ReplEnum = MapNames::findReplacedName(EnumConstantRule::EnumNamesMap,
-                                                ECD->getName().str());
-    requestHelperFeatureForEnumNames(ECD->getName().str(), DRE);
-    auto ItEnum = EnumConstantRule::EnumNamesMap.find(ECD->getName().str());
+                                                RefString);
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/ExprAnalysis.cpp:437]: Make sure enum with no enum type name can be matched

Comment (AndyCHHuang) [inline clang/lib/DPCT/ExprAnalysis.cpp:437]: Make sure enum with no enum type name can be matched

File: clang/lib/DPCT/ExprAnalysis.h [hunk 3]
```diff
@@ -204,6 +204,10 @@ class ExprAnalysis {
     initSourceRange(TL.getSourceRange());
     analyzeType(TL);
   }
+  inline void analyze(const TemplateSpecializationTypeLoc &TSTL) {
+    initSourceRange(TSTL.getSourceRange());
+    analyzeType(TSTL);
+  }
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/ExprAnalysis.h]: remove

Comment (AndyCHHuang) [inline clang/lib/DPCT/ExprAnalysis.h]: remove

File: clang/test/dpct/cuda_atomic.cu [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// RUN: dpct -out-root %T/cuda_atomic %s --cuda-include-path="%cuda-path/include" -- -x cuda --cuda-host-only -fno-delayed-template-parsing
+// RUN: FileCheck %s --match-full-lines --input-file %T/cuda_atomic/cuda_atomic.dp.cpp
+
+#include <cuda/atomic>
+
+int main(){
+    // CHECK: cuda::atomic_ext<int, sycl::memory_order::relaxed, sycl::memory_scope::system> a;
```
Comment (AndyCHHuang) [inline clang/test/dpct/cuda_atomic.cu]: should be no cuda

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  };
+}
+
+std::function<std::string(const TypeLoc *)>
+makeTemplateArgCreator(std::string ArgStr) {
+  return [=](const TypeLoc *TL) -> std::string {
+    return ArgStr;
+  };
+}
+
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/TypeLocRewriters.cpp]: keep one of them, rename

File: clang/lib/DPCT/TypeLocRewriters.h [hunk 6]
```diff
@@ -0,0 +1,126 @@
+#include "CallExprRewriter.h"
+
+namespace clang {
+namespace dpct {
+
+template <class TypeNameT, class... TemplateArgsT> class TypeLocPrinter {
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/TypeLocRewriters.h]: Reuse existing printer

File: clang/lib/DPCT/TypeLocRewriters.h [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    dpct::print(Stream, TypeName);
+    Stream << "<";
+    TAs.print(Stream);
+    Stream << ">";
+  }
+};
+
+class TypeLocRewriter {
+protected:
+  const TypeLoc *TL;
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/TypeLocRewriters.h]: use class instead of pointer

File: clang/lib/DPCT/TypeLocRewriters.h [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  StringRef OriginalTypeName;
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/TypeLocRewriters.h]: Remove

**Review by zhiweij1 — APPROVED** [review]

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  bool operator()(const TypeLoc TL) {
+    if(auto TSTL = TL.getAs<TemplateSpecializationTypeLoc>()){
+      return TSTL.getNumArgs() == Count;
+    }
+    return false;
+  }
+};
+
+template <class TypeNameT, class... TemplateArgsT>
+std::shared_ptr<TypeLocRewriterFactoryBase> creatTypeLocRewriterFactory(
```
Comment (ziranzha) [inline clang/lib/DPCT/TypeLocRewriters.cpp]: shoulde be 'createTypeLocRewriterFactory'

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 10]
```diff
@@ -17357,6 +17357,28 @@ void ComplexAPIRule::runRule(
...
+    const ast_matchers::MatchFinder::MatchResult &Result) {
+  if (auto TL = getNodeAsType<TypeLoc>(Result, "loc")) {
+    ExprAnalysis EA;
+    EA.analyze(*TL);
+    emplaceTransformation(EA.getReplacement());
+    EA.applyAllSubExprRepl();
+  }
+}
+
+REGISTER_RULE(TemplateSpecializationTypeLocRule)
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp:17257]: the new rule should be recorded into MigrationRules.inc.

File: clang/lib/DPCT/ExprAnalysis.h [hunk 11]
```diff
@@ -388,8 +395,9 @@ class ExprAnalysis {
   }
 
   template <class T> void analyzeTemplateSpecializationType(const T &TL) {
-    for (size_t i = 0; i < TL.getNumArgs(); ++i)
+    for (size_t i = 0; i < TL.getNumArgs(); ++i){
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ExprAnalysis.h]: Please format this line L398. there should be space before {.

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 12]
```diff
@@ -0,0 +1,84 @@
+#include "TypeLocRewriters.h"
```
Comment (zhimingwang36) [inline clang/lib/DPCT/TypeLocRewriters.cpp:9]: For new added file, add  license header.

File: clang/lib/DPCT/TypeLocRewriters.h [hunk 13]
```diff
@@ -0,0 +1,105 @@
+#include "CallExprRewriter.h"
```
Comment (zhimingwang36) [inline clang/lib/DPCT/TypeLocRewriters.h:9]: Please add  license header for new added file.

File: clang/lib/DPCT/APINamesTemplateType.inc [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+TYPE_REWRITE_ENTRY(
+    "cuda::atomic",
+    TYPE_CONDITIONAL_FACTORY(CheckTemplateArgCount(2),
+                             TYPE_FACTORY(STR("atomic_ext"), TEMPLATE_ARG(0),
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APINamesTemplateType.inc]: use dpct::atomic

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**